### PR TITLE
Fix: Percy CLI hangs on Node.js 26 during Chromium extraction — adm-zip variant (PER-10062)

### PR DESCRIPTION
**Pull request by prklm10**
Fixes PER-10062. **Alternative implementation to #2343** — same root cause and test contract, different extraction library. Pick one.

## Root cause
Identical to #2343: `extract-zip@2.0.1` → `yauzl@2` relies on undefined stream `destroy()` behavior that Node.js 26 tightened, so `openReadStream()` callbacks never fire on deflate entries and Chromium extraction hangs forever (https://github.com/thejoshwolfe/yauzl/issues/176). `percy exec` blocks before spawning the wrapped command → no tests run, builds finalize `no_snapshots`.

## Fix
Replaces `extract-zip` with a local `packages/core/src/unzip.js` built on **`adm-zip@^0.6.0`**, which inflates synchronously and is unaffected by the Node 26 stream change.

**Why not adm-zip's own `extractAllTo`:** verified on Node 26 that it writes symlink entries as regular files containing the target path — this breaks the `Chromium.app` bundle on macOS (framework symlinks) — and mishandles windows-style directory attributes. The wrapper iterates `getEntries()` and restores file modes, directories, and symlinks itself, with a zip-slip guard (reachable here, unlike with yauzl 3 which pre-validates entry names).

## Trade-offs vs #2343 (yauzl@3)
- adm-zip is sync/buffered: each entry is inflated fully in memory (largest Chromium entry is a few hundred MB) vs yauzl's streaming. Fine for install-time use; worth knowing on 4 GB CI containers.
- adm-zip has no transitive deps; yauzl@3 pulls only `pend`.
- `getData()` CRC-validates every entry (corrupted data throws instead of writing bad files).

## Testing
- Same 10-spec `unzip.test.js` contract as #2343 (hand-crafted zip fixtures: stored + deflated entries, nested dirs, `__MACOSX/` skip, zip-slip rejection, default + executable modes, symlinks, windows dir attrs, corrupted data): **pass on Node 14.21.3 and Node 26.5.0**, 100% coverage on the new module.
- All 51 install + unzip specs pass under Node 14 (CI env).
- Includes the same master coverage repairs as #2343 (canonicalHost spec + pragma) so CI can go green.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Review by ninadbstack — APPROVED** [review]

File: packages/core/test/unit/unzip.test.js [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+  it('rejects entries that would escape the target directory', async () => {
+    fs.writeFileSync(archive, makeZip([
+      { name: '../evil.txt', data: 'oops' }
+    ]));
+
+    // yauzl 3 rejects relative paths at parse time; the unzip module also
+    // guards against zip-slip for entries yauzl allows through
+    await expectAsync(unzip(archive, { dir: path.join(tmp, 'out') }))
+      .toBeRejectedWithError(/invalid relative path|Out of bound path/);
```
Comment (prklm10) [inline packages/core/test/unit/unzip.test.js:135]: **[Medium] Stale comment + dead regex branch**

The comment says "yauzl 3 rejects relative paths at parse time," but this module no longer uses yauzl at all — it's leftover from the extract-zip era. The `invalid relative path` alternative in the regex is dead (adm-zip never throws that string).

**Suggestion:** describe the zip-slip guard instead and tighten the regex to `/Out of bound path/`.

_Reviewer: stack:code-reviewer_

File: packages/cli-exec/test/start.test.js [hunk 2]
```diff
@@ -12,6 +12,21 @@ describe('percy exec:start', () => {
     return promise;
   }
 
+  // `start(...)` resolves on process exit, not on listen, so poll the
+  // healthcheck endpoint until the server is reachable. A fixed sleep proved
+  // flaky on slow Windows CI runners where startup can exceed a second and
+  // `request` only retries ECONNREFUSED 5 times at 50ms intervals.
+  async function waitForHealthcheck(port = 5338, deadline = Date.now() + 15000) {
```
Comment (prklm10) [inline packages/cli-exec/test/start.test.js:19]: **[Medium] Poll deadline exceeds Jasmine's non-Windows timeout**

The 15s default deadline is longer than `jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000` set for non-Windows platforms in `scripts/test-helpers.js` (25s is Windows-only). A slow Linux/macOS runner needing >10s would fail with Jasmine's generic "Async function did not complete within 10000ms" instead of the informative healthcheck error this helper is meant to provide.

**Suggestion:** cap the default deadline below 10s on non-Windows, or raise the affected specs' own timeout to at least the poll deadline.

_Reviewer: stack:code-reviewer_

File: packages/core/src/unzip.js [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+// stat mode constants for entry external file attributes
+const IFMT = 0o170000;
+const IFDIR = 0o040000;
+const IFLNK = 0o120000;
+
+// Returns the file mode of an extracted entry, with defaults matching the
+// previous extract-zip behavior when the archive does not provide one
+function extractedMode(mode, isDir) {
+  return (mode || (isDir ? 0o755 : 0o644)) & 0o777;
```
Comment (prklm10) [inline packages/core/src/unzip.js:13]: **[Low] `mode || default` conflates "no mode" with "mode 0"**

An entry whose archive mode is genuinely `0` silently gets the 0755/0644 default instead of 0. Harmless for Chromium archives, but this module isn't Chromium-specific.

**Suggestion:** add a one-line comment noting the tradeoff.

_Reviewer: stack:code-reviewer_

**Review by prklm10 — COMMENTED** [review]
Claude Code Review (automated) — 5 inline finding(s). Full report in the PR comment below. Verdict: Passed.

File: packages/core/src/unzip.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+    if (isDir) {
+      await fs.promises.mkdir(dest, { recursive: true, mode: extractedMode(mode, true) });
+    } else {
+      await fs.promises.mkdir(path.dirname(dest), { recursive: true });
+      // synchronous inflate with crc validation; throws on corrupted data
+      let contents = entry.getData();
+
+      if (isSymlink) {
+        await fs.promises.symlink(contents.toString('utf8'), dest);
```
Comment (prklm10) [inline packages/core/src/unzip.js:60]: **[Low] Symlink creation isn't idempotent on retried extractions**

`fs.promises.symlink` throws `EEXIST` if `dest` already exists. `download()` in `install.js` only skips extraction when the final executable exists, so a retry after a mid-extraction failure into the same revision outdir could fail permanently until `.local-chromium/<revision>` is deleted manually.

**Suggestion:** `await fs.promises.rm(dest, { force: true })` before `symlink`, or remove the outdir on extraction failure.

_Reviewer: stack:code-reviewer_

File: packages/cli-exec/test/start.test.js [hunk 5]
```diff
@@ -12,6 +12,21 @@ describe('percy exec:start', () => {
...
+    while (true) {
+      try {
+        return await request(`http://localhost:${port}/percy/healthcheck`, { retries: 0, noProxy: true });
+      } catch (err) {
```
Comment (prklm10) [inline packages/cli-exec/test/start.test.js:23]: **[Low] Retries on any error, not just connection-refused**

If the server binds but the healthcheck route itself errors, the loop masks the real error for up to 15s before surfacing it, delaying failure feedback.

**Suggestion:** optionally retry only connection errors (e.g. `err.code === 'ECONNREFUSED'`) and rethrow others immediately.

_Reviewer: stack:code-reviewer_

**Comment by prklm10** [comment]
# Claude Code PR Review

**PR:** https://github.com/percy/cli/pull/2344  •  **Head:** 31d304b  •  **Reviewers:** stack:code-reviewer

## Summary
Replaces the unmaintained `extract-zip` (yauzl@2) with a local adm-zip-based `packages/core/src/unzip.js` to fix the Node 26 Chromium-extraction deadlock (PER-10062), preserving file modes, directories, and symlinks with a zip-slip guard; also de-flakes the Windows `percy exec:start` CI specs by polling the healthcheck endpoint instead of a fixed 1s sleep.

## Review Table

| Priority | Category | Check | Status | Notes |
|----------|----------|-------|--------|-------|
| High | Security | No hardcoded secrets or credentials | Pass | Only test-fixture placeholder tokens (`<<PERCY_TOKEN>>`) |
| High | Security | Authentication/authorization checks present | N/A | No auth surface in this change |
| High | Security | Input validation and sanitization | Pass | Absolute-dir check, `realpath`, and zip-slip guard in `unzip.js`; CRC validation on inflate (full security lens disabled for this run — handled separately) |
| High | Security | No IDOR — resource ownership validated | N/A | No resource access by identifier |
| High | Security | No SQL injection (parameterized queries) | N/A | No database code |
| High | Correctness | Logic is correct, handles edge cases | Pass | Mode/symlink/dir handling verified against adm-zip source; per-file parent `mkdir` is archive-order independent; minor non-blocking nits noted below (#3, #4) |
| High | Correctness | Error handling is explicit, no swallowed exceptions | Pass | Corrupted data throws (CRC), out-of-bound entries throw, relative target dirs rejected |
| High | Correctness | No race conditions or concurrency issues | Pass | PR removes a race (fixed-sleep test wait → deterministic healthcheck polling) |
| Medium | Testing | New code has corresponding tests | Pass | 10-spec `unzip.test.js` with hand-crafted zip fixtures; 100% coverage on the new module |
| Medium | Testing | Error paths and edge cases tested | Fail | #2: 15s poll deadline exceeds the 10s Jasmine timeout on non-Windows — slow POSIX runner would fail with a generic timeout instead of the informative error |
| Medium | Testing | Existing tests still pass (no regressions) | Pass | All 46 CI checks green on head, including Windows |
| Medium | Performance | No N+1 queries or unbounded data fetching | Pass | #5 noted: adm-zip inflates synchronously (event-loop blocking) — acceptable for a one-shot install step, worth knowing |
| Medium | Performance | Long-running tasks use background jobs | N/A | CLI install step, inherently foreground |
| Medium | Quality | Follows existing codebase patterns | Pass | Matches core module style; esm loader mocks updated consistently across install specs |
| Medium | Quality | Changes are focused (single concern) | Pass | Extraction swap + the CI de-flake needed to land it; no scope creep |
| Low | Quality | Meaningful names, no dead code | Pass | Dead regex branch flagged in #1 (test-only) |
| Low | Quality | Comments explain why, not what | Fail | #1: stale "yauzl 3 rejects relative paths" comment in `unzip.test.js` — leftover from the extract-zip era; module under test no longer uses yauzl |
| Low | Quality | No unnecessary dependencies added | Pass | Swaps 7 transitive packages (yauzl/fd-slicer/pend/buffer-crc32/get-stream@5/pump/extract-zip) for zero-dependency `adm-zip@0.6.0` |

## Findings

- **File:** `packages/core/test/unit/unzip.test.js:132`
- **Severity:** Medium
- **Reviewer:** stack:code-reviewer
- **Issue:** Stale comment claims "yauzl 3 rejects relative paths at parse time," but this module no longer uses yauzl at all; the `invalid relative path` alternative in the assertion regex is dead (adm-zip never throws that string).
- **Suggestion:** Update the comment to describe the zip-slip guard and tighten the regex to `/Out of bound path/`.

- **File:** `packages/cli-exec/test/start.test.js:19`
- **Severity:** Medium
- **Reviewer:** stack:code-reviewer
- **Issue:** `waitForHealthcheck` defaults to a 15s deadline, but `scripts/test-helpers.js` sets `jasmine.DEFAULT_TIMEOUT_INTERVAL = 10000` on non-Windows (25000 only on win32). A slow Linux/macOS runner needing >10s would fail with Jasmine's generic "Async function did not complete within 10000ms" instead of the informative healthcheck error.
- **Suggestion:** Cap the default deadline below 10s on non-Windows, or raise the affected specs' own timeout to ≥ the poll deadline.

- **File:** `packages/core/src/unzip.js:13`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** `extractedMode` uses `mode || default`, conflating "archive provides no mode" with "mode explicitly 0" — a real mode of 0 silently becomes 0755/0644. Harmless for Chromium archives, but the module is not Chromium-specific.
- **Suggestion:** Add a one-line comment noting the tradeoff (or check for `undefined` explicitly if archives with intentional 0 modes ever matter).

- **File:** `packages/core/src/unzip.js:60`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** `fs.promises.symlink` throws `EEXIST` if `dest` already exists. `download()` in `install.js` only skips extraction when the final executable exists, so a retry after a mid-extraction failure into the same revision-named outdir could fail permanently until `.local-chromium/<revision>` is manually deleted.
- **Suggestion:** `await fs.promises.rm(dest, { force: true })` before `symlink`, or remove `outdir` on extraction failure.

- **File:** `packages/core/src/unzip.js` (module-level)
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** Extraction is now fully synchronous (adm-zip `readFileSync` + sync inflate) — for a ~150–250MB Chromium zip this blocks the event loop for the duration, a behavioral shift from streaming extract-zip. Reasonable (and what sidesteps the Node 26 deadlock) for a one-shot install step.
- **Suggestion:** Keep the sync/buffered tradeoff documented in the PR description (already covered there) so future profiling isn't surprised.

- **File:** `packages/cli-exec/test/start.test.js:22`
- **Severity:** Low
- **Reviewer:** stack:code-reviewer
- **Issue:** `waitForHealthcheck` retries on any error, not just connection-refused — if the healthcheck route itself errors, the loop masks it for up to 15s before surfacing, delaying failure feedback.
- **Suggestion:** Optionally retry only on connection errors (`err.code === 'ECONNREFUSED'` etc.) and rethrow others immediately.

---

**Verdict: PASS** — no Critical/High findings; the extraction logic was verified against adm-zip's source, coverage is thorough, and all 46 CI checks pass. The two Medium test-hygiene findings are worth a quick follow-up but are non-blocking.
